pkg/oidc/verify.go

```diff
@@ -0,0 +1,74 @@
+package oidc
+
+import (
+	"context"
+	"fmt"
+	"github.com/coreos/go-oidc"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/logger"
+)
+
+type IDTokenVerifier struct {
+	*oidc.IDTokenVerifier
+	*IDTokenVerificationOptions
+}
+
+type IDTokenVerificationOptions struct {
+	AudienceClaim  string
+	ClientID       string
+	ExtraAudiences []string
+}
+
+func NewVerifier(iv *oidc.IDTokenVerifier, vo *IDTokenVerificationOptions) *IDTokenVerifier {
+	return &IDTokenVerifier{iv, vo}
+}
+
+func (v *IDTokenVerifier) Verify(ctx context.Context, rawIDToken string) (*oidc.IDToken, error) {
+	token, err := v.IDTokenVerifier.Verify(ctx, rawIDToken)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse default id_token claims: %v", err)
+	}
+
+	claims := map[string]interface{}{}
+	if err := token.Claims(&claims); err != nil {
+		return nil, fmt.Errorf("failed to parse default id_token claims: %v", err)
+	}
+
+	if isValidAudience, err := v.verifyAudience(claims); !isValidAudience {
+		return nil, err
+	}
+
+	// maybye other additional validation/verification for future purposes...
+
+	return token, err
+}
+
+func (v *IDTokenVerifier) verifyAudience(claims map[string]interface{}) (bool, error) {
+	allowedAudiences := append([]string{v.ClientID}, v.IDTokenVerificationOptions.ExtraAudiences...)
```

Comment:
Since this doesn't change for the lifetime of the verifier, what about constructing this at the time the verifier is constructed? You could also make it a map to make the lookup easier